lib/ops/opsservice/agents.go

```diff
@@ -442,47 +442,32 @@ func (r *AgentPeerStore) authenticatePeer(token string) (*storage.ProvisioningTo
 }
 
 func (r *AgentPeerStore) validatePeer(ctx context.Context, group *agentGroup, info storage.System,
-	req pb.PeerJoinRequest, clusterName string) error {
+	req pb.PeerJoinRequest, token storage.ProvisioningToken) error {
 	if group.hasPeer(req.Addr, info.GetHostname()) {
 		return nil
 	}
 
-	if err := r.checkHostname(ctx, group, req.Addr, info.GetHostname(), clusterName); err != nil {
+	if err := r.checkHostname(ctx, group, req.Addr, info.GetHostname(), token); err != nil {
 		return trace.Wrap(err)
 	}
 
-	if err := r.checkLicense(ctx, int(group.NumPeers()), clusterName, info); err != nil {
+	if err := r.checkLicense(ctx, int(group.NumPeers()), token.SiteDomain, info); err != nil {
 		return trace.Wrap(err)
 	}
 
 	return nil
 }
 
-func (r *AgentPeerStore) checkHostname(ctx context.Context, group *agentGroup, addr, hostname, clusterName string) error {
-	// collect hostnames from existing servers (for expand)
-	servers, err := r.teleport.GetServers(ctx, clusterName, nil)
-	if err != nil && !trace.IsNotFound(err) {
-		return trace.Wrap(err)
-	}
-
-	var existingServers []string
-	for _, server := range servers {
-		hostname := server.GetLabels()[ops.Hostname]
-		if hostname == "" {
-			log.Warnf("Server hostname is empty: %+v.", server)
-			continue
+func (r *AgentPeerStore) checkHostname(ctx context.Context, group *agentGroup, addr, hostname string, token storage.ProvisioningToken) error {
+	if err := r.isPartOfActiveOperation(addr, token); err != nil {
+		r.WithError(err).Warn("Failed to check whether the server is part of the active operation.")
```

Comment:
Nit: This will say "failed to check" even if the check actually succeeded but returned NotFound (which is an expected scenario).